pkg/readiness/pruner/pruner_test.go

```diff
@@ -0,0 +1,270 @@
+			WatchManger:      wm,
+			ControllerSwitch: sw,
+			Tracker:          tracker,
+			ProcessExcluder:  process.Get(),
+			MutationSystem:   mutationSystem,
+			ExpansionSystem:  expansion.NewSystem(mutationSystem),
+			ProviderCache:    frameworksexternaldata.NewCache(),
+			CacheMgr:         cm,
+			SyncEventsCh:     events,
+		}
+		require.NoError(t, controller.AddToManager(mgr, &opts), "registering controllers")
+	}
+
+	ep := &ExpectationsPruner{
+		cacheMgr: cm,
+		tracker:  tracker,
+	}
+	if o.addExpectationsPruner {
+		require.NoError(t, mgr.Add(ep), "adding expectationspruner as a runnable")
+	}
+
+	return &testResources{expectationsPruner: ep, manager: mgr, k8sClient: c}
+}
+
+// Test_ExpectationsMgr_DeletedSyncSets tests scenarios in which SyncSet and Config resources
+// get deleted after tracker expectations have been populated and we need to reconcile
+// the GVKs that are in the data client (via the cachemaanger) and the GVKs that are expected
+// by the Tracker.
+func Test_ExpectationsMgr_DeletedSyncSets(t *testing.T) {
+	tts := []struct {
+		name             string
+		fixturesPath     string
+		syncsetsToDelete []string
+		deleteConfig     string
+		// not starting controllers approximates missing events in the informers cache
+		startControllers bool
+	}{
+		{
+			name:             "delete all syncsets",
+			fixturesPath:     "testdata/syncsets-overlapping",
+			syncsetsToDelete: []string{"syncset-1", "syncset-2", "syncset-3"},
+		},
+		{
+			name:             "delete syncs and configs",
+			fixturesPath:     "testdata/syncsets-config-disjoint",
+			syncsetsToDelete: []string{"syncset-1"},
+			deleteConfig:     "config",
+		},
+		{
+			name:             "delete one syncset",
+			fixturesPath:     "testdata/syncsets-resources",
+			syncsetsToDelete: []string{"syncset-2"},
+			startControllers: true,
+		},
+	}
+
+	for _, tt := range tts {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancelFunc := context.WithCancel(context.Background())
+
+			require.NoError(t, testutils.ApplyFixtures(tt.fixturesPath, cfg), "applying base fixtures")
+
+			testRes := setupTest(ctx, t, testOptions{addConstrollers: tt.startControllers})
+
+			testutils.StartManager(ctx, t, testRes.manager)
+
+			require.Eventually(t, func() bool {
+				return testRes.expectationsPruner.tracker.Populated()
+			}, timeout, tick, "waiting on tracker to populate")
+
+			for _, name := range tt.syncsetsToDelete {
+				u := &unstructured.Unstructured{}
+				u.SetGroupVersionKind(syncsetGVK)
+				u.SetName(name)
+
+				require.NoError(t, testRes.k8sClient.Delete(ctx, u), fmt.Sprintf("deleting syncset %s", name))
+			}
+			if tt.deleteConfig != "" {
+				u := &unstructured.Unstructured{}
+				u.SetGroupVersionKind(configGVK)
+				u.SetNamespace("gatekeeper-system")
+				u.SetName(tt.deleteConfig)
+
+				require.NoError(t, testRes.k8sClient.Delete(ctx, u), fmt.Sprintf("deleting config %s", tt.deleteConfig))
+			}
+
+			testRes.expectationsPruner.pruneNotWatchedGVKs()
```

Comment:
What guarantees that there are expectations to prune here? For the first two tests we don't start controllers, so there is nothing to call "Observe", which helps, but what about for the third test?

IMO the third case is sufficiently covered by the `missedInformers` test below, and can be removed to sidestep this issue.